apps/language_server/lib/language_server/providers/references.ex

```diff
@@ -1,96 +1,47 @@
     end)
   end
 
   def supported? do
     Mix.Tasks.Xref.__info__(:functions) |> Enum.member?({:calls, 0})
   end
 
-  defp xref_at_cursor(text, line, character) do
-    env_at_cursor = line_environment(text, line)
-    %{aliases: aliases} = env_at_cursor
-
-    subject_at_cursor(text, line, character)
-    # TODO: Don't call into here directly
-    |> ElixirSense.Core.Source.split_module_and_func(aliases)
-    |> expand_mod_fun(env_at_cursor)
-    |> add_arity(env_at_cursor)
-    |> callers()
-  end
-
-  defp line_environment(text, line) do
-    # TODO: Don't call into here directly
-    ElixirSense.Core.Parser.parse_string(text, true, true, line + 1)
-    |> ElixirSense.Core.Metadata.get_env(line + 1)
-  end
-
-  defp subject_at_cursor(text, line, character) do
-    # TODO: Don't call into here directly
-    ElixirSense.Core.Source.subject(text, line + 1, character + 1)
-  end
-
-  defp expand_mod_fun({nil, nil}, _environment), do: nil
-
-  defp expand_mod_fun(mod_fun, %{imports: imports, aliases: aliases, module: module}) do
-    # TODO: Don't call into here directly
-    mod_fun = ElixirSense.Core.Introspection.actual_mod_fun(mod_fun, imports, aliases, module)
-
-    case mod_fun do
-      {mod, nil} -> {mod, nil}
-      {mod, fun} -> {mod, fun}
-    end
-  end
-
-  defp add_arity({mod, fun}, %{scope: {fun, arity}, module: mod}), do: {mod, fun, arity}
-  defp add_arity({mod, fun}, _env), do: {mod, fun, nil}
-
-  defp callers(mfa) do
-    if Mix.Project.umbrella?() do
-      umbrella_calls()
-    else
-      Mix.Tasks.Xref.calls()
-    end
-    |> Enum.filter(caller_filter(mfa))
-  end
-
-  def umbrella_calls() do
-    build_dir = Path.expand(Mix.Project.config()[:build_path])
-
-    Mix.Project.apps_paths()
-    |> Enum.flat_map(fn {app, path} ->
-      Mix.Project.in_project(app, path, [build_path: build_dir], fn _ ->
-        Mix.Tasks.Xref.calls()
-        |> Enum.map(fn %{file: file} = call ->
-          Map.put(call, :file, Path.expand(file))
-        end)
-      end)
-    end)
+  defp build_reference(ref) do
+    %{
+      range: %{
+        start: %{line: ref.range.start.line, column: ref.range.start.column},
+        end: %{line: ref.range.end.line, column: ref.range.end.column}
+      },
+      uri: ref.uri
+    }
   end
 
-  defp caller_filter({module, nil, nil}), do: &match?(%{callee: {^module, _, _}}, &1)
-  defp caller_filter({module, func, nil}), do: &match?(%{callee: {^module, ^func, _}}, &1)
-  defp caller_filter({module, func, arity}), do: &match?(%{callee: {^module, ^func, ^arity}}, &1)
+  defp build_loc(reference) do
+    # Adjust for ElixirSense 1-based indexing
+    line_start = reference.range.start.line - 1
+    line_end = reference.range.end.line - 1
+    column_start = reference.range.start.column - 1
+    column_end = reference.range.end.column - 1
 
-  defp build_location(%{file: file, line: line}) do
     %{
-      "uri" => SourceFile.path_to_uri(file),
+      "uri" => SourceFile.path_to_uri(reference.uri),
```

Comment:
For references to variables in the current text file `uri` will be `nil`. Will it work?

Reply:
Ah, good catch. I'm able to reproduce that locally:

```
06:59:52.672 [error] Process #PID<0.1010.0> raised an exception
** (FunctionClauseError) no function clause matching in IO.chardata_to_string/1
    (elixir) lib/io.ex:557: IO.chardata_to_string(nil)
    (elixir) lib/path.ex:680: Path.expand_home/1
    (elixir) lib/path.ex:154: Path.expand/1
    (language_server) lib/language_server/source_file.ex:54: ElixirLS.LanguageServer.SourceFile.path_to_uri/1
    (language_server) lib/language_server/providers/references.ex:41: ElixirLS.LanguageServer.Providers.References.build_loc/1
    (elixir) lib/enum.ex:1336: Enum."-map/2-lists^map/1-0-"/2
    (kernel) global.erl:425: :global.trans/4
    (language_server) lib/language_server/server.ex:355: anonymous fn/5 in ElixirLS.LanguageServer.Server.handle_request/2
```

@lukaszsamson if the `reference.uri` returned from ElixirSense is `nil` can we assume that the uri is the current file? If not could we change ElixirSense to return `:nofile` (or similar) to make the case more explicit?

Reply:
I'm afraid we cannot assume that. IIRC `nil` will be returned also for macro generated fun calls, builtins etc. I think we should return either `:nofile` or `:current_buffer` to clearly discriminate those cases.

Reply:
Yeah `:current_buffer` is the case that we really need. Where `:current_buffer` means the reference was found in the `text` that was passed in.

Based on your "Edit" are you saying that we can make an assumption? (or perhaps I'm reading that incorrectly)

Reply:
@lukaszsamson do you have an example that would generate a `:nofile` response? I've tried with a macro generated function but I just get back an empty list from `ElixirSense.references/3` when hovering over the `macro_a_func()` call with these modules:

```
defmodule ElixirLS.Test.MacroA do
  defmacro __using__(_) do
    quote do
      def macro_a_func do
        :ok
      end
    end
  end
end
```

```
defmodule ElixirLS.Test.UsesMacroA do
  use ElixirLS.Test.MacroA

  def my_fun do
    macro_a_func()
  end

  def my_other_fun do
    macro_a_func()
  end
end
```